File: server/services/arxiv.ts
import { XMLParser } from 'fast-xml-parser';

export interface ArxivEntry {
  id: string;
  title: string;
  summary: string;
  author: any;
  published: string;
  category: any;
  link: any;
}

// Raw paper data from ArXiv (before AI summarization)
export interface ArxivPaper {
  id: string;
  title: string;
  authors: string;
  abstract: string;
  arxivUrl: string;
  publishedDate: string;
  categories: string[];
}

export class ArxivService {
  private baseUrl = 'http://export.arxiv.org/api/query';
  private parser = new XMLParser({
    ignoreAttributes: false,
    attributeNamePrefix: "@_"
  });

  async searchPapers(topic: string, keywords?: string, subdomain?: string, days: number = 7): Promise<ArxivPaper[]> {
    try {
      // Construct search query
      let searchQuery = `all:${topic}`;
      
      if (keywords) {
        const keywordArray = keywords.split(',').map(k => k.trim());
        searchQuery += ` AND (${keywordArray.map(k => `all:${k}`).join(' OR ')})`;
      }

      if (subdomain && subdomain !== "all") {
        searchQuery += ` AND cat:${subdomain}`;
      } else {
        // Default to CS categories
        searchQuery += ` AND (cat:cs.AI OR cat:cs.LG OR cat:cs.CL OR cat:cs.CV OR cat:cs.RO OR cat:cs.NE)`;
      }

      // Get papers from last N days
      const cutoffDate = new Date();
      cutoffDate.setDate(cutoffDate.getDate() - days);
      const dateFilter = cutoffDate.toISOString().split('T')[0].replace(/-/g, '');

      const params = new URLSearchParams({
        search_query: searchQuery,
        start: '0',
        max_results: '50',
        sortBy: 'submittedDate',
        sortOrder: 'descending'
      });

      const response = await fetch(`${this.baseUrl}?${params}`);
      
      if (!response.ok) {
        throw new Error(`ArXiv API error: ${response.status} ${response.statusText}`);
      }

      const xmlData = await response.text();
      const parsed = this.parser.parse(xmlData);
      
      const feed = parsed.feed;
      if (!feed || !feed.entry) {
        return [];
      }

      // Handle both single entry and array of entries
      const entries: ArxivEntry[] = Array.isArray(feed.entry) ? feed.entry : [feed.entry];
      
      // Filter by date and convert to our ArxivPaper format
      const papers: ArxivPaper[] = entries
        .filter(entry => {
          const publishedDate = new Date(entry.published);
          return publishedDate >= cutoffDate;
        })
        .map(entry => this.convertToPaper(entry))
        .slice(0, 10); // Return up to 10 papers for weekly digest

      return papers;
    } catch (error) {
      console.error('Error fetching from ArXiv:', error);
      throw new Error('Failed to fetch papers from ArXiv');
    }
  }

  private convertToPaper(entry: ArxivEntry): ArxivPaper {
    // Handle authors - can be single object or array
    let authors = '';
    if (Array.isArray(entry.author)) {
      authors = entry.author.map(a => a.name).join(', ');
    } else if (entry.author) {
      authors = entry.author.name || 'Unknown';
    }

    // Handle categories
    let categories: string[] = [];
    if (Array.isArray(entry.category)) {
      categories = entry.category.map(c => c["@_term"]);
    } else if (entry.category) {
      categories = [entry.category["@_term"]];
    }

    // Extract arXiv ID from the ID URL
    const arxivId = entry.id.split('/').pop() || entry.id;
    
    return {
      id: arxivId,
      title: entry.title.replace(/\n/g, ' ').trim(),
      authors: authors,
      abstract: entry.summary.replace(/\n/g, ' ').trim(),
      arxivUrl: `https://arxiv.org/abs/${arxivId}`,
      publishedDate: entry.published,
      categories: categories,
    };
  }
}

File: backend/services/arxiv.py
import httpx
import xml.etree.ElementTree as ET
from datetime import datetime, timedelta
from typing import List, Optional
from dataclasses import dataclass


@dataclass
class ArxivPaper:
    """Raw paper data from ArXiv."""
    id: str
    title: str
    authors: str
    abstract: str
    arxiv_url: str
    published_date: str
    categories: List[str]


class ArxivService:
    """Service for fetching papers from ArXiv API."""

    BASE_URL = "http://export.arxiv.org/api/query"

    async def search_papers(
        self,
        topic: str,
        keywords: Optional[str] = None,
        days: int = 7
    ) -> List[ArxivPaper]:
        """Search ArXiv for papers matching the topic."""
        try:
            # Build search query
            search_query = f"all:{topic}"

            if keywords:
                keyword_list = [k.strip() for k in keywords.split(",")]
                keyword_query = " OR ".join([f"all:{k}" for k in keyword_list])
                search_query += f" AND ({keyword_query})"

            # Default to CS categories
            search_query += " AND (cat:cs.AI OR cat:cs.LG OR cat:cs.CL OR cat:cs.CV OR cat:cs.RO OR cat:cs.NE)"

            # Calculate cutoff date
            cutoff_date = datetime.now() - timedelta(days=days)

            params = {
                "search_query": search_query,
                "start": "0",
                "max_results": "50",
                "sortBy": "submittedDate",
                "sortOrder": "descending"
            }

            async with httpx.AsyncClient() as client:
                response = await client.get(self.BASE_URL, params=params, timeout=30.0)
                response.raise_for_status()

            # Parse XML response
            papers = self._parse_response(response.text, cutoff_date)

            # Return up to 10 papers
            return papers[:10]

        except Exception as e:
            print(f"[ArxivService] Error fetching papers: {e}")
            return []

    def _parse_response(self, xml_text: str, cutoff_date: datetime) -> List[ArxivPaper]:
        """Parse ArXiv XML response."""
        papers = []

        # Define namespaces
        namespaces = {
            "atom": "http://www.w3.org/2005/Atom",
            "arxiv": "http://arxiv.org/schemas/atom"
        }

        root = ET.fromstring(xml_text)

        for entry in root.findall("atom:entry", namespaces):
            try:
                # Parse published date
                published_str = entry.find("atom:published", namespaces).text
                published_date = datetime.fromisoformat(published_str.replace("Z", "+00:00"))

                # Filter by date
                if published_date.replace(tzinfo=None) < cutoff_date:
                    continue

                # Extract ID from URL
                id_url = entry.find("atom:id", namespaces).text
                arxiv_id = id_url.split("/")[-1]

                # Get title
                title = entry.find("atom:title", namespaces).text
                title = " ".join(title.split())  # Normalize whitespace

                # Get authors
                authors = []
                for author in entry.findall("atom:author", namespaces):
                    name = author.find("atom:name", namespaces).text
                    authors.append(name)
                authors_str = ", ".join(authors)

                # Get abstract
                abstract = entry.find("atom:summary", namespaces).text
                abstract = " ".join(abstract.split())  # Normalize whitespace

                # Get categories
                categories = []
                for category in entry.findall("atom:category", namespaces):
                    term = category.get("term")
                    if term:
                        categories.append(term)

                paper = ArxivPaper(
                    id=arxiv_id,
                    title=title,
                    authors=authors_str,
                    abstract=abstract,
                    arxiv_url=f"https://arxiv.org/abs/{arxiv_id}",
                    published_date=published_str,
                    categories=categories
                )
                papers.append(paper)

            except Exception as e:
                print(f"[ArxivService] Error parsing entry: {e}")
                continue

        return papers


# Singleton instance
arxiv_service = ArxivService()
